src/test/java/com/microsoft/aad/msal4j/TokenRequestTest.java

```diff
@@ -101,13 +80,80 @@ public void executeOAuthRequest_SCBadRequestErrorInteractionRequired_ClaimsChall
 
         try {
             request.executeOauthRequestAndProcessResponse();
-            Assert.fail("Expected ClaimsChallengeException was not thrown");
-        } catch (ClaimsChallengeException ex) {
+            Assert.fail("Expected AuthenticationServiceException was not thrown");
+        } catch (AuthenticationServiceException ex) {
             Assert.assertEquals(claims.replace("\\", ""), ex.claims());
         }
         PowerMock.verifyAll();
     }
 
+    @Test
+    public void executeOAuthRequest_SCBadRequestErrorInvalidGrant_SubErrorFiltered()
+            throws SerializeException, ParseException, AuthenticationException,
+            IOException, URISyntaxException {
+
+        TokenRequest request = createMockedTokenRequest();
+
+        OAuthHttpRequest msalOAuthHttpRequest = PowerMock.createMock(OAuthHttpRequest.class);
+
+        HTTPResponse httpResponse = new HTTPResponse(HTTPResponse.SC_BAD_REQUEST);
+
+        String claims = "{\\\"access_token\\\":{\\\"polids\\\":{\\\"essential\\\":true,\\\"values\\\":[\\\"5ce770ea-8690-4747-aa73-c5b3cd509cd4\\\"]}}}";
+
+        String content = "{\"error\":\"invalid_grant\"," +
+                "\"error_description\":\"AADSTS65001: description\\r\\nCorrelation ID: 3a...5a\\r\\nTimestamp:2017-07-15 02:35:05Z\"," +
+                "\"error_codes\":[50076]," +
+                "\"timestamp\":\"2017-07-15 02:35:05Z\"," +
+                "\"trace_id\":\"0788...000\"," +
+                "\"correlation_id\":\"3a...95a\"," +
+                "\"suberror\":\"client_mismatch\"," +
+                "\"claims\":\"" + claims + "\"}";
+        httpResponse.setContent(content);
+        httpResponse.setContentType(CommonContentTypes.APPLICATION_JSON);
+
+        EasyMock.expect(request.toOauthHttpRequest()).andReturn(msalOAuthHttpRequest).times(1);
+        EasyMock.expect(msalOAuthHttpRequest.send()).andReturn(httpResponse).times(1);
+
+        PowerMock.replay(request, msalOAuthHttpRequest);
+
+        try {
+            request.executeOauthRequestAndProcessResponse();
+            Assert.fail("Expected AuthenticationServiceException was not thrown");
+        } catch (AuthenticationServiceException ex) {
+            Assert.assertEquals(claims.replace("\\", ""), ex.claims());
+            Assert.assertEquals(ex.subError(), "");
```

Comment:
>Assert.assertEquals(ex.subError(), ""); [](start = 12, length = 39)

I dont' see any tests really validating the sub error. Also the subError is not what should be exposed, the exposed property is Classification and it will only be a subset of the suberrors. Please work with Bogdan on these details.